File: frontend/src/components/benchmark/content/dfs/implementations/dfs.cpp
#include <vector>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

using namespace std;

class DFS {
private:
    unordered_map<int, unordered_set<int>> adjacency;
    unordered_set<int> nodes;

    void dfs_recursive(int node, unordered_set<int>& visited, vector<int>& order) {
        visited.insert(node);
        order.push_back(node);
        
        for (int neighbor : adjacency[node]) {
            if (visited.find(neighbor) == visited.end()) {
                dfs_recursive(neighbor, visited, order);
            }
        }
    }

public:
    void add_edge(int u, int v) {
        adjacency[u].insert(v);
        adjacency[v].insert(u);
        nodes.insert(u);
        nodes.insert(v);
    }

    void add_directed_edge(int u, int v) {
        adjacency[u].insert(v);
        nodes.insert(u);
        nodes.insert(v);
    }

    void build_from_edges(const vector<pair<int, int>>& edges, bool directed = false) {
        for (const auto& edge : edges) {
            if (directed) {
                add_directed_edge(edge.first, edge.second);
            } else {
                add_edge(edge.first, edge.second);
            }
        }
    }

    vector<int> dfs_recursive(int start) {
        vector<int> order;
        unordered_set<int> visited;
        dfs_recursive(start, visited, order);
        return order;
    }

    vector<int> dfs_iterative(int start) {
        vector<int> order;
        unordered_set<int> visited;
        stack<int> st;
        
        st.push(start);
        
        while (!st.empty()) {
            int node = st.top();
            st.pop();
            
            if (visited.find(node) != visited.end()) continue;
            
            visited.insert(node);
            order.push_back(node);
            
            for (int neighbor : adjacency[node]) {
                if (visited.find(neighbor) == visited.end()) {
                    st.push(neighbor);
                }
            }
        }
        
        return order;
    }

    vector<int> dfs_path(int start, int goal) {
        unordered_map<int, int> parent;
        unordered_set<int> visited;
        stack<int> st;
        
        parent[start] = -1;
        st.push(start);
        
        while (!st.empty()) {
            int node = st.top();
            st.pop();
            
            if (visited.find(node) != visited.end()) continue;
            visited.insert(node);
            
            if (node == goal) {
                vector<int> path;
                while (node != -1) {
                    path.push_back(node);
                    node = parent[node];
                }
                reverse(path.begin(), path.end());
                return path;
            }
            
            for (int neighbor : adjacency[node]) {
                if (visited.find(neighbor) == visited.end()) {
                    parent[neighbor] = node;
                    st.push(neighbor);
                }
            }
        }
        
        return {};
    }

    vector<vector<int>> find_components() {
        vector<vector<int>> components;
        unordered_set<int> visited;
        
        for (int node : nodes) {
            if (visited.find(node) == visited.end()) {
                vector<int> component;
                dfs_recursive(node, visited, component);
                components.push_back(component);
            }
        }
        
        return components;
    }

    bool is_connected() {
        if (nodes.empty()) return true;
        
        int start = *nodes.begin();
        unordered_set<int> visited;
        vector<int> order;
        dfs_recursive(start, visited, order);
        
        return visited.size() == nodes.size();
    }

    bool has_cycle() {
        unordered_set<int> visited;
        unordered_set<int> rec_stack;
        
        for (int node : nodes) {
            if (visited.find(node) == visited.end()) {
                if (has_cycle_util(node, visited, rec_stack, -1)) {
                    return true;
                }
            }
        }
        return false;
    }

    bool has_cycle_util(int node, unordered_set<int>& visited, 
                        unordered_set<int>& rec_stack, int parent) {
        visited.insert(node);
        rec_stack.insert(node);
        
        for (int neighbor : adjacency[node]) {
            if (rec_stack.find(neighbor) != rec_stack.end() && neighbor != parent) {
                return true;
            }
            
            if (visited.find(neighbor) == visited.end()) {
                if (has_cycle_util(neighbor, visited, rec_stack, node)) {
                    return true;
                }
            }
        }
        
        rec_stack.erase(node);
        return false;
    }
};

File: frontend/src/components/benchmark/content/dfs/implementations/dfs.go
package main

type DFS struct {
    adjacency map[int]map[int]bool
    nodes     map[int]bool
}

func NewDFS() *DFS {
    return &DFS{
        adjacency: make(map[int]map[int]bool),
        nodes:     make(map[int]bool),
    }
}

func (d *DFS) AddEdge(u, v int) {
    if d.adjacency[u] == nil {
        d.adjacency[u] = make(map[int]bool)
    }
    if d.adjacency[v] == nil {
        d.adjacency[v] = make(map[int]bool)
    }
    d.adjacency[u][v] = true
    d.adjacency[v][u] = true
    d.nodes[u] = true
    d.nodes[v] = true
}

func (d *DFS) AddDirectedEdge(u, v int) {
    if d.adjacency[u] == nil {
        d.adjacency[u] = make(map[int]bool)
    }
    if d.adjacency[v] == nil {
        d.adjacency[v] = make(map[int]bool)
    }
    d.adjacency[u][v] = true
    d.nodes[u] = true
    d.nodes[v] = true
}

func (d *DFS) BuildFromEdges(edges [][2]int, directed bool) {
    for _, edge := range edges {
        if directed {
            d.AddDirectedEdge(edge[0], edge[1])
        } else {
            d.AddEdge(edge[0], edge[1])
        }
    }
}

func (d *DFS) dfsRecursive(node int, visited map[int]bool, order *[]int) {
    visited[node] = true
    *order = append(*order, node)

    for neighbor := range d.adjacency[node] {
        if !visited[neighbor] {
            d.dfsRecursive(neighbor, visited, order)
        }
    }
}

func (d *DFS) DfsRecursive(start int) []int {
    order := []int{}
    visited := make(map[int]bool)
    d.dfsRecursive(start, visited, &order)
    return order
}

func (d *DFS) DfsIterative(start int) []int {
    order := []int{}
    visited := make(map[int]bool)
    stack := []int{start}

    for len(stack) > 0 {
        node := stack[len(stack)-1]
        stack = stack[:len(stack)-1]

        if visited[node] {
            continue
        }

        visited[node] = true
        order = append(order, node)

        for neighbor := range d.adjacency[node] {
            if !visited[neighbor] {
                stack = append(stack, neighbor)
            }
        }
    }

    return order
}

func (d *DFS) DfsPath(start, goal int) []int {
    parent := make(map[int]int)
    visited := make(map[int]bool)
    stack := []int{start}
    parent[start] = -1

    for len(stack) > 0 {
        node := stack[len(stack)-1]
        stack = stack[:len(stack)-1]

        if visited[node] {
            continue
        }
        visited[node] = true

        if node == goal {
            path := []int{}
            for node != -1 {
                path = append([]int{node}, path...)
                node = parent[node]
            }
            return path
        }

        for neighbor := range d.adjacency[node] {
            if !visited[neighbor] {
                parent[neighbor] = node
                stack = append(stack, neighbor)
            }
        }
    }

    return []int{}
}

func (d *DFS) FindComponents() [][]int {
    components := [][]int{}
    visited := make(map[int]bool)

    for node := range d.nodes {
        if !visited[node] {
            component := []int{}
            d.dfsRecursive(node, visited, &component)
            components = append(components, component)
        }
    }

    return components
}

func (d *DFS) IsConnected() bool {
    if len(d.nodes) == 0 {
        return true
    }

    var start int
    for node := range d.nodes {
        start = node
        break
    }

    visited := make(map[int]bool)
    order := []int{}
    d.dfsRecursive(start, visited, &order)

    return len(visited) == len(d.nodes)
}

func (d *DFS) HasCycle() bool {
    visited := make(map[int]bool)
    recStack := make(map[int]bool)

    for node := range d.nodes {
        if !visited[node] {
            if d.hasCycleUtil(node, visited, recStack, -1) {
                return true
            }
        }
    }
    return false
}

func (d *DFS) hasCycleUtil(node int, visited, recStack map[int]bool, parent int) bool {
    visited[node] = true
    recStack[node] = true

    for neighbor := range d.adjacency[node] {
        if recStack[neighbor] && neighbor != parent {
            return true
        }

        if !visited[neighbor] {
            if d.hasCycleUtil(neighbor, visited, recStack, node) {
                return true
            }
        }
    }

    recStack[node] = false
    return false
}